javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -104,9 +132,11 @@ private static int treeSize(int branchCount, int depthShift) {
         } else if (n >= length()) {
             return empty();
         } else {
-            final NodeModifier drop = Arrays::copyDrop;
-            final Object[] root = modifyLeaf(array, depthShift, offset + n, drop, drop);
-            return new BitMappedTrie<>(root, offset + n, length() - n, depthShift);
+            final int index = offset + n;
+            final Object[] root = arePointingToSameLeaf(0, n)
+                                  ? array
+                                  : modifyLeaf(array, depthShift, index, Arrays::copyDrop, Identity);
+            return new BitMappedTrie<T>(root, index, length() - n, depthShift).collapseHeight();
```

Comment:
`collapseHeight()` is always called directly after constructing the BMT - which is a smell :)

We can save one BMT instance (in the worst case) by changing `collapseHeight()` to be a static factory method:

```java
/* drop root node while it has a single element */
private static BitMappedTrie<T> collapseHeight(Object[] array, int offset, int length, int shift) {
    for (; shift > 0; shift -= BRANCHING_BASE) {
        final int skippedElements = array.length - 1;
        if (skippedElements != digit(offset, shift)) {
            break;
        }
        array = getAt(array, skippedElements);
        offset -= treeSize(skippedElements, shift);
    }
    return new BitMappedTrie<>(array, offset, length, shift);
}
```

Then we replace

```java
return new BitMappedTrie<T>(root, index, length() - n, depthShift).collapseHeight();
```

by

```java
return collapseHeight(root, index, length() - n, depthShift);
```

Maybe we should rename `collapseHeight(...)` then to `of(...)`? But not necessarily.